packages/client/lib/miner/miner.ts

```diff
@@ -0,0 +1,206 @@
+  }
+
+  /**
+   * Sets the next block assembly to latestBlock.timestamp + period
+   */
+  private async chainUpdated() {
+    const latestBlock = await this.latestBlock()
+    const target = latestBlock.header.timestamp.muln(1000).addn(this.period).sub(new BN(Date.now()))
+    const timeout = BN.max(new BN(0), target).toNumber()
+    this.config.logger.debug(
+      `Miner: Chain updated with block ${latestBlock.header.number.toNumber()}. Queuing next block assembly in ${Math.round(
+        timeout / 1000
+      )}s`
+    )
+    this.queueNextAssembly(timeout)
+  }
+
+  /**
+   * Start miner
+   */
+  start(): boolean {
+    if (!this.config.mine || this.running) {
+      return false
+    }
+    this.running = true
+    if (this.config.chainCommon.consensusType() === ConsensusType.ProofOfAuthority) {
+      this.period = this.config.chainCommon.consensusConfig().period * 1000
+    }
+    this.config.events.on(Event.CHAIN_UPDATED, this.chainUpdated.bind(this))
+    this.queueNextAssembly()
+    this.config.logger.info(`Miner started. Assembling next block in ${this.period / 1000}s`)
+    return true
+  }
+
+  /**
+   * Assembles a block from txs in the TxPool and adds it to the chain.
+   * If a new block is received while assembling it will abort.
+   */
+  async assembleBlock() {
+    if (this.assembling) {
+      return
+    }
+    this.assembling = true
+
+    // Abort if a new block is received while assembling this block
+    let interrupt = false
+    const setInterrupt = () => {
+      interrupt = true
+    }
+    this.config.events.on(Event.CHAIN_UPDATED, setInterrupt)
+
+    const parentBlock = await this.latestBlock()
+    const { gasLimit } = parentBlock.header
+    const [signerAddress, signerPrivKey] = this.config.accounts[0]
+
+    // Use a copy of the vm to not modify the existing state.
+    // The state will be updated when the newly assembled block
+    // is inserted into the canonical chain.
+    const vmCopy = this.synchronizer.execution.vm.copy()
+
+    // If we are on the genesis block the canonical
+    // genesis state has not been initialiezd yet.
+    if (parentBlock.header.number.isZero()) {
+      await vmCopy.stateManager.generateCanonicalGenesis()
+    }
+
+    let difficulty
+    if (this.config.chainCommon.consensusType() === ConsensusType.ProofOfAuthority) {
+      // Determine if signer is INTURN (2) or NOTURN (1)
+      const signers = vmCopy.blockchain.cliqueActiveSigners()
+      const signerIndex = signers.findIndex((address) => address.equals(signerAddress))
+      const inTurn = parentBlock.header.number.addn(1).modn(signers.length) === signerIndex
+      difficulty = inTurn ? 2 : 1
+    }
+
+    const blockBuilder = await vmCopy.buildBlock({
+      parentBlock,
+      headerData: {
+        difficulty,
+        gasLimit,
+      },
+      blockOpts: {
+        cliqueSigner: signerPrivKey,
+      },
+    })
+
+    let baseFee
+    if (this.config.chainCommon.isActivatedEIP(1559)) {
```

Comment:
One nit/half-nit: this is actually one off on the transition block, right? 🤔

Not sure if it's dramatically important in this context.

Reply:
hm, well we can't run `calcNextBaseFee()` unless the common already has 1559 activated or it will [throw](https://github.com/ethereumjs/ethereumjs-monorepo/blob/5b91ccde820dae0b4eb969eb53d88227c9582931/packages/block/src/header.ts#L665).

Reply:
Ah, I am getting what you mean, but I think we still have got some inconsistencies around this whole base fee intitialization thing and HF switching on the block builder, need to give this some short thought.

So the one thing is, that I guess that you are missing the `initialBaseFee` case here and - with our current design - you might need to replicate somewhat the code from [here](https://github.com/ethereumjs/ethereumjs-monorepo/blob/master/packages/block/src/header.ts#L643) in the Block header if I get this right?

I do have to say that I find this whole design a bit unfortunate though. With this introduction of the [BlockHeader.calcNextBaseFee()](https://github.com/ethereumjs/ethereumjs-monorepo/blob/master/packages/block/src/header.ts#L663) function we have introduced some functionality which is referring to the HF state of the next block which often leads to such a "one-off" factor.

Might it improve the situation if we add a function `getNextCommon()` (or `getNextHeaderCommon()`)  to the `Block` library? This function would do a copy of the internal `Common` instance, set the HF with the `hardforkByBlockNumber` option to the HF of `block.number + 1` and return the Common instance.

We could use this on several fronts:

1. Internally in `calcNextBaseFee()` and then we could return the initial base fee for block with a block number one below the 1559 HF (this should only be called if 1559 is not already activated since otherwise we would restrict on non-blocknumber-based HF setting use cases).
2. You can use this here instead of adding the initial base fee logic manually.

Not 100% sure on this suggestion and still on the 80% level or so, might need some sleeping over.

A bit independent from the suggestion above. When I see this correctly I think the block builder is also (and generally) one off with the HF when creating a new block. this takes `this._vm._common` as a [Common input for the next block](https://github.com/ethereumjs/ethereumjs-monorepo/blob/master/packages/vm/src/buildBlock.ts#L67) and on a HF transition this will be one-off potentially leading to undefined behavior.

This might rather be a case for the `hardforkByBlockNumber` though and we should generally - as well - be careful here to not restrict on block-number-independent use cases (as Hardhat is eventually (likely?) using this for. So `hardforkByBlockNumber` should (I guess) rather be an additional block builder option passed through to the block and initially set to `false`.

Does all/most of the above make sense? 😛

Reply:
> So the one thing is, that I guess that you are missing the `initialBaseFee` case here and - with our current design - you might need to replicate somewhat the code from [here](https://github.com/ethereumjs/ethereumjs-monorepo/blob/master/packages/block/src/header.ts#L643) in the Block header if I get this right?

yes good point, will make sure that is handled.

> I do have to say that I find this whole design a bit unfortunate though (around `calcNextBaseFee` and the checks and the like). With this introduction of the [BlockHeader.calcNextBaseFee()](https://github.com/ethereumjs/ethereumjs-monorepo/blob/master/packages/block/src/header.ts#L663) function we have introduced some functionality which is referring to the HF state of the next block which often leads to such a "one-off" factor.
> 
> Might it improve the situation if we add a function `getNextCommon()` (or `getNextHeaderCommon()`) to the `Block` library? This function would do a copy of the internal `Common` instance, set the HF with the `hardforkByBlockNumber` option to the HF of `block.number + 1` and return the Common instance.
> 
> We could use this on several fronts:
> 
> 1. Internally in `calcNextBaseFee()` and then we could return the initial base fee for block with a block number one below the 1559 HF (this should only be called if 1559 is not already activated since otherwise we would restrict on non-blocknumber-based HF setting use cases).
> 2. You can use this here instead of adding the initial base fee logic manually.
> 
> Not 100% sure on this suggestion and still on the 80% level or so, might need some sleeping over.

yeah this is a little complex, seems like a pretty good suggestion though so I'll keep thinking about it and possibly try an implementation in this PR if it helps simplify/improve things.

> A bit independent from the suggestion above. When I see this correctly I think the block builder is also (and generally) one off with the HF when creating a new block. this takes `this._vm._common` as a [Common input for the next block](https://github.com/ethereumjs/ethereumjs-monorepo/blob/master/packages/vm/src/buildBlock.ts#L67) and on a HF transition this will be one-off potentially leading to undefined behavior.
> 
> This might rather be a case for the `hardforkByBlockNumber` though and we should generally - as well - be careful here to not restrict on block-number-independent use cases (as Hardhat is eventually (likely?) using this for. So `hardforkByBlockNumber` should (I guess) rather be an additional block builder option passed through to the block and initially set to `false`.

This can already be passed in to `blockOpts` and defaults to `false`, so that should already be the existing behavior. For the miner I should probably pass this option as `true` to satisfy some of these edge cases.